Configure ping in HTTP2

aik-jahoda opened the pull request:
Allows configure HTTP2 ping on SocketsHttpHandler:

```c#
namespace System.Net.Http
{
    public class SocketsHttpHandler
    {
        public TimeSpan KeepAlivePingDelay { get; set; }
        public TimeSpan KeepAlivePingTimeout { get; set; }
        public HttpKeepAlivePingPolicy KeepAlivePingPolicy { get; set; }
    }
    public enum HttpKeepAlivePingPolicy
    {
        WithActiveRequests,
        Always
    }
}
```

The `KeepAlivePingDelay` defines how often the ping is sent. Any incoming frame can postpone the ping request. It means during vivacious communication the ping will not be sent at all as the communication itself manifest the connection is live. The time resolution is 1s.

`KeepAlivePingTimeout` defines the ping ack timeout. The time resolution is 1s.


`KeepAlivePingPolicy` defines in what circumstances the ping is sent:
1. `WithActiveRequests` - only if the connection has an active stream
2. `Always` - send the ping always - even if the connection is returned to the connection pool and no streams are active.

When the ping detects the connection is active, it closes the connection.

What to expect
-----
- The connection is closed as soon as possible, not during Send attempt
- A proxy will not close connection with active frame flow

What to not expect
-----
- Connection latency - the API don't measure connection latency


fixes #31198

Commit:
init

--amend

Commit:
Http2 ping

Commit:
Merge remote-tracking branch 'upstream/master' into jajahoda/httpping

Commit:
clean up

aik-jahoda commented on src/libraries/System.Net.Http.WinHttpHandler/ref/System.Net.Http.WinHttpHandler.cs:
```diff
@@ -48,4 +48,10 @@ public partial class WinHttpHandler : System.Net.Http.HttpMessageHandler
         protected override System.Threading.Tasks.Task<System.Net.Http.HttpResponseMessage> SendAsync(System.Net.Http.HttpRequestMessage request, System.Threading.CancellationToken cancellationToken) { throw null; }
     }
+
+    public enum HttpKeepAlivePingPolicy
```
This is weird. This enum is there as it is used in HttpHandlerDefaults. 
Only way how to prevent it what I have in my mind is make HttpHandlerDefaults partial and move SocketsHttpHandler related properties to System.Net.Http only.

ManickaP replied:
Definitely split it, this shouldn't be in `WinHttpHandler` at all.

aik-jahoda replied:
done

Commit:
enable tests

Commit:
Fix mac build

Commit:
Merge branch 'master' of github.com:dotnet/runtime into jajahoda/httpping

Commit:
Split HttpHandlerDefaults

Commit:
Merge branch 'master' of github.com:dotnet/runtime into jajahoda/httpping

Commit:
Merge remote-tracking branch 'origin/jajahoda/httpping' into jajahoda/httpping

Commit:
Fix mac build

Commit:
Fix browser and test csproj

ManickaP commented on src/libraries/System.Net.Http/tests/FunctionalTests/SocketsHttpHandlerTest.cs:
```diff
@@ -279,6 +279,123 @@ public sealed class SocketsHttpHandler_HttpClientHandler_ServerCertificates_Test
+                        //_output.WriteLine("ack sent");
+                    }
+                    else
+                        await Assert.ThrowsAsync<OperationCanceledException>(() => connection.ReadPingAsync(pingTimeout));
```
NIT: Could you please enclose the `else` block in braces since the `if` one is.

ManickaP commented on src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/SocketsHttpHandler.cs:
```diff
@@ -277,6 +277,49 @@ public TimeSpan Expect100ContinueTimeout
             }
         }
 
+        public TimeSpan KeepAlivePingDelay
```
I'm interested in what a documentation for these properties will be. What is the magical combination to turn the PING sending off?

scalablecory replied:
Yes, these properties require XML comments. Setting this to `Timeout.InfiniteTimeSpan` should turn it off.

JamesNK replied:
I've already written XML docs in ASP.NET Core. They might be a helpful starting point.

https://github.com/dotnet/aspnetcore/blob/29ceed2b2a6bcd9f3064f51cc78858e22c8544c4/src/Servers/Kestrel/Core/src/Http2Limits.cs#L149-L158

aik-jahoda replied:
Kestrel comments reused 👍

ManickaP reviewed (commented):
Two major questions:
- Is the timer with `HeartBeat` right approach?
- Is it worth the extra class `Http2KeepAlive`?

@scalablecory you're opinions?

scalablecory commented on src/libraries/System.Net.Http/src/System/Net/Http/HttpHandlerDefaults.cs:
```diff
@@ -0,0 +1,17 @@
+    internal static partial class HttpHandlerDefaults
+    {
+        public static readonly TimeSpan DefaultKeepAlivePingTimeout = TimeSpan.FromSeconds(20);
+        public static readonly TimeSpan DefaultKeepAlivePingDelay = TimeSpan.FromSeconds(0);
```
We should use `TimeSpan.InfiniteTimeout` as the "off" mode.

That said, I feel like it'd be fine to have this default to on. @JamesNK do you know what a reasonable default delay is? would 60 seconds be too long?

JamesNK replied:
Java and golang default to off. CCore defaults to 5 minutes.

I think we agreed to have it off by default. I'm happy leaving it off. That avoids the overhead of maintaining the timer and related functionality unless someone opts in.

JamesNK replied:
By the way, `KeepAlivePingDelay` and `KeepAlivePingTimeout` are nullable in ASP.NET Core. null equals off.

If you don't want nullable then I agree that `TimeSpan.InfiniteTimeout` should equal off.

aik-jahoda replied:
The default changed to `Timeout.InfiniteTimeSpan`

scalablecory commented on src/libraries/System.Net.Http/src/System/Net/Http/HttpHandlerDefaults.cs:
```diff
@@ -0,0 +1,17 @@
+    /// </summary>
+    internal static partial class HttpHandlerDefaults
+    {
+        public static readonly TimeSpan DefaultKeepAlivePingTimeout = TimeSpan.FromSeconds(20);
```
JIT will optimize this better if you do:

```suggestion
        public static TimeSpan DefaultKeepAlivePingTimeout => TimeSpan.FromTicks(TimeSpan.TicksPerSecond * 20);
```

aik-jahoda replied:
Why the property? To make it lazy?

scalablecory replied:
the readonly field will take up memory and require initialization.
the property will be optimized 100% into a constant `mov rax, 1234h; ret`.

ManickaP replied:
@scalablecory It comes from other [default timeouts](https://github.com/dotnet/runtime/pull/40257/files/20f44cf98a4de5734deb1e1d146cbb891c2a9cd0#diff-1ef6a69559626aa1f87616cd3f0106c1R17) defined on the class (but in a different file due to being linked in WinHttpHandler).

If it gets changed here, it might cause confusion why is here used different approach.

aik-jahoda replied:
I'll add the change to the rest values

scalablecory commented on src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/Http2KeepAlive.cs:
```diff
@@ -0,0 +1,99 @@
+            PingSent
+        }
+
+        internal class Http2KeepAlive
```
I would not put this in a separate class; it'd be better to bake it all into `Http2Connection`.

JamesNK replied:
In ASP.NET Core I put it in a separate class to reduce server connection memory usage when keep alive is not enabled. I don't know if that is a consideration here.

scalablecory replied:
Since this is per-connection and each connection handles 100+ requests, I'm not _super_ concerned by resource usage of a handful of fields.

aik-jahoda replied:
the extra class is to separate the logic which is related only to keep alive and not for the rest of connection.

aik-jahoda replied:
the class was removed

scalablecory commented on src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/HttpConnectionPoolManager.cs:
```diff
@@ -130,6 +132,15 @@ public HttpConnectionPoolManager(HttpConnectionSettings settings)
                 }
             }
+
+            _heartBeatTimer = new Timer(static state =>
```
This needs to be constructed similar to the other timer, in a way which avoids capturing the execution context:

https://github.com/dotnet/runtime/blob/c0707135c4969be520efae02c9593d481c57cc8f/src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/HttpConnectionPoolManager.cs#L92-L121

scalablecory replied:
The code above is already doing this; perhaps this can be done within the same try/finally. It should also reuse the `WeakReference` instance.

aik-jahoda replied:
I have overlook the context. Fixed.

scalablecory commented on src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/Http2KeepAlive.cs:
```diff
@@ -0,0 +1,99 @@
+                {
+                    case KeepAliveState.None:
+                        // Check whether keep alive delay has passed since last frame received
+                        if (_keepAlivePingDelay > TimeSpan.Zero && now > _nextPingRequestTimestamp)
```
```suggestion
                        if (_keepAlivePingDelay != Timeout.InfiniteTimeSpan && now > _nextPingRequestTimestamp)
```

aik-jahoda replied:
The delay should be greater than 0 to enable ping logic. As Timeout.InfiniteTimeSpan is negative, it is already covered.

aik-jahoda replied:
I changed the KeepAlivePingDelay property default to Timeout.InfiniteTimeSpan

scalablecory commented on src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/Http2KeepAlive.cs:
```diff
@@ -0,0 +1,99 @@
+                            // Set the status directly to ping sent and set the timestamp
+                            _state = KeepAliveState.PingSent;
+                            _pingTimeoutTimestamp = now.Add(_keepAlivePingTimeout);
+                            Interlocked.Increment(ref _pingPayload);
+                            _connection.SendPingAsync(_pingPayload);
```
Reading `_pingPayload` can tear. Use the return from the increment instead.

Also, you are ignoring the returned task from `SendPingAsync`. If that task fails, it should probably cause the connection to abort.

```suggestion
                            long pingPayload = Interlocked.Increment(ref _pingPayload);
                            _connection.SendPingAsync(pingPayload);
```

scalablecory commented on src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/HttpConnectionPool.cs:
```diff
@@ -1689,6 +1689,17 @@ private static bool GetIsWindows7Or2008R2()
 
+        internal void HeartBeat()
+        {
+            Http2Connection[]? localHttp2Connections = _http2Connections;
```
This variable is modified concurrently.

```suggestion
            Http2Connection[]? localHttp2Connections = Volatile.Read(ref _http2Connections);
```

aik-jahoda replied:
Pointer read is atomic, isn't it? _http2Connections is already volatile, so memory barrier should be in place.

scalablecory commented on src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/HttpConnectionPool.cs:
```diff
@@ -1689,6 +1689,17 @@ private static bool GetIsWindows7Or2008R2()
+            Http2Connection[]? localHttp2Connections = _http2Connections;
+            if (localHttp2Connections != null)
+            {
+                foreach (var http2Connection in localHttp2Connections)
```
Do not use `var` when not immediately obvious from RHS.

```suggestion
                foreach (Http2Connection http2Connection in localHttp2Connections)
```

scalablecory commented on src/libraries/System.Net.Http/tests/FunctionalTests/SocketsHttpHandlerTest.cs:
```diff
@@ -279,6 +279,123 @@ public sealed class SocketsHttpHandler_HttpClientHandler_ServerCertificates_Test
+                },
+                async server =>
+                {
+                    Http2LoopbackConnection connection = await server.EstablishConnectionAsync();
```
```suggestion
                    using Http2LoopbackConnection connection = await server.EstablishConnectionAsync();
```

scalablecory commented on src/libraries/System.Net.Http/tests/FunctionalTests/SocketsHttpHandlerTest.cs:
```diff
@@ -279,6 +279,123 @@ public sealed class SocketsHttpHandler_HttpClientHandler_ServerCertificates_Test
+        [Fact]
+        public void DefaultValues()
+        {
+            using (var handler = new SocketsHttpHandler())
+            {
+                Assert.Equal(TimeSpan.FromSeconds(20), handler.KeepAlivePingTimeout);
+                Assert.Equal(TimeSpan.FromSeconds(0), handler.KeepAlivePingDelay);
+                Assert.Equal(HttpKeepAlivePingPolicy.Always, handler.KeepAlivePingPolicy);
+            }
```
Prefer a using declaration when the scope is obvious, to avoid excessive indentation.

```suggestion
            using var handler = new SocketsHttpHandler();
            Assert.Equal(TimeSpan.FromSeconds(20), handler.KeepAlivePingTimeout);
            Assert.Equal(TimeSpan.FromSeconds(0), handler.KeepAlivePingDelay);
            Assert.Equal(HttpKeepAlivePingPolicy.Always, handler.KeepAlivePingPolicy);
```

scalablecory commented on src/libraries/System.Net.Http/tests/FunctionalTests/SocketsHttpHandlerTest.cs:
```diff
@@ -279,6 +279,123 @@ public sealed class SocketsHttpHandler_HttpClientHandler_ServerCertificates_Test
+                    };
+                    handler.SslOptions.RemoteCertificateValidationCallback = (_, __, ___, ____) => true;
+
+                    using (HttpClient client = new HttpClient(handler))
```
a using declaration can be used here too.

scalablecory reviewed (changes requested):
First pass review done; will provide a 2nd pass once comments are resolved.

ManickaP commented:
Just thinking, but instead of the `HeartBeat` timer, could we just make a method with a loop with `await Task.Delay(PingDelay)`? Similar to `ProcessIncomingFrames`, i.e. started but not awaited:
https://github.com/dotnet/runtime/blob/master/src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/Http2Connection.cs#L158

However, I don't know if this would be worse or better on the resources than the shared timer across all connections.

JamesNK commented on src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/SocketsHttpHandler.cs:
```diff
@@ -277,6 +277,49 @@ public TimeSpan Expect100ContinueTimeout
+            get => _settings._keepAlivePingDelay;
+            set
+            {
+                if ((value < TimeSpan.Zero && value != Timeout.InfiniteTimeSpan) ||
```
Timeout.InfiniteTimeSpan should equal off

Zero shouldn't be allowed. A keep alive ping of zero doesn't make sense.

In ASP.NET Core the minimum allow value is 1 second. Partly we did that because there is a global server heartbeat of once a second and that is the lowest resolution allowed, but also because pinging more often than once a second doesn't make sense. Why would someone want to set `KeepAlivePingDelay = TimeSpan.FromMilliseconds(1)`?

scalablecory replied:
I don't know if it makes to automatically clip the value for the user, but if we did then something like 10 seconds seems very reasonably more frequent than anyone would want to use.

aik-jahoda replied:
0 and Timeout.InfiniteTimeSpan means we don't ping at all. It is default behavior for SocketsHttpHandler and the ping should be enabled explicitly by this property

aik-jahoda replied:
The default was changed to `Timeout.InfiniteTimeSpan` and 0 is forbidden now

JamesNK commented on src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/SocketsHttpHandler.cs:
```diff
@@ -277,6 +277,49 @@ public TimeSpan Expect100ContinueTimeout
+            }
+        }
+
+        public TimeSpan KeepAlivePingTimeout
```
Starting point for docs:

https://github.com/dotnet/aspnetcore/blob/29ceed2b2a6bcd9f3064f51cc78858e22c8544c4/src/Servers/Kestrel/Core/src/Http2Limits.cs#L174-L183

JamesNK commented on src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/SocketsHttpHandler.cs:
```diff
@@ -277,6 +277,49 @@ public TimeSpan Expect100ContinueTimeout
+            get => _settings._keepAlivePingTimeout;
+            set
+            {
+                if ((value < TimeSpan.Zero && value != Timeout.InfiniteTimeSpan) ||
```
Consider a minimum time of 1 second?

JamesNK commented on src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/Http2Connection.cs:
```diff
@@ -119,6 +122,8 @@ public Http2Connection(HttpConnectionPool pool, Connection connection)
             _pendingWindowUpdate = 0;
             _idleSinceTickCount = Environment.TickCount64;
 
+            _keepAlive = new Http2KeepAlive(this, _pool.Settings);
```
If keep alive is not enabled, i.e. KeepAlivePingDelay is off, then you don't need to create this type.

JamesNK commented on src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/HttpConnectionPoolManager.cs:
```diff
@@ -130,6 +132,15 @@ public HttpConnectionPoolManager(HttpConnectionSettings settings)
+                {
+                    thisRef.HeartBeat();
+                }
+            }, new WeakReference<HttpConnectionPoolManager>(this), TimeSpan.FromSeconds(1), TimeSpan.FromSeconds(1));
```
All implementations I have seen of pinging says that unreasonably low values for ping delay and timeout are not allowed.

1 second for the heartbeat is fine - this is what we do in Kestrel - but I think the KeepAlivePingDelay and KeepAlivePingTimeout should disallow values smaller than a second. That way developers won't expect to be able to configure `KeepAlivePingDelay = TimeSpan.FromMillisecond(1)` and expect a ping to be sent every millisecond.

JamesNK commented on src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/HttpConnectionPoolManager.cs:
```diff
@@ -130,6 +132,15 @@ public HttpConnectionPoolManager(HttpConnectionSettings settings)
                 }
             }
+
+            _heartBeatTimer = new Timer(static state =>
```
If there is no ping configured then there is no need for this timer. Only create when pings are enabled?

JamesNK commented on src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/Http2KeepAlive.cs:
```diff
@@ -0,0 +1,99 @@
+                        }
+                        break;
+                    case KeepAliveState.PingSent:
+                        if (_keepAlivePingTimeout != TimeSpan.MaxValue)
```
I think this might be copy pasta from ASP.NET Core 😉

aik-jahoda replied:
MaxValue means we don't care about timeout at all. Anyway the _pingTimeoutTimestamp would be so huge that I can remove the condition and handle it as normal huge timeout.

JamesNK commented:
> Just thinking, but instead of the `HeartBeat` timer, could we just make a method with a loop with `await Task.Delay(PingDelay)`? Similar to `ProcessIncomingFrames`, i.e. started but not awaited:
> [`master`/src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/Http2Connection.cs#L158](https://github.com/dotnet/runtime/blob/master/src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/Http2Connection.cs?rgh-link-date=2020-08-04T15%3A32%3A44Z#L158)
> 
> However, I don't know if this would be worse or better on the resources than the shared timer across all connections.

One problem is time would slowly drift. Each iteration would be PingDelay + time to execute heartbeat logic. You'd need to calculate that time and subtract it from PingDelay when doing the next delay. Even with that time would probably slowly drift.

I haven't looked at the Timer source but it will no doubt have optimizations that make it a better choice than Task.Delay calls in a loop.

scalablecory commented:
>> Just thinking, but instead of the `HeartBeat` timer, could we just make a method with a loop with `await Task.Delay(PingDelay)`? Similar to `ProcessIncomingFrames`, i.e. started but not awaited:
>> https://github.com/dotnet/runtime/blob/master/src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/Http2Connection.cs#L158
> 
>> However, I don't know if this would be worse or better on the resources than the shared timer across all connections.
>
> I haven't looked at the Timer source but it will no doubt have optimizations that make it a better choice than Task.Delay calls in a loop.

Using `Timer` here is a lot more efficient. `Timer` has no allocations once it is going. `Task.Delay` allocates a new timer for every call.

ManickaP commented:
> > > Just thinking, but instead of the `HeartBeat` timer, could we just make a method with a loop with `await Task.Delay(PingDelay)`? Similar to `ProcessIncomingFrames`, i.e. started but not awaited:
> > > https://github.com/dotnet/runtime/blob/master/src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/Http2Connection.cs#L158
> > 
> > 
> > > However, I don't know if this would be worse or better on the resources than the shared timer across all connections.
> > 
> > 
> > I haven't looked at the Timer source but it will no doubt have optimizations that make it a better choice than Task.Delay calls in a loop.
> 
> Using `Timer` here is a lot more efficient. `Timer` has no allocations once it is going. `Task.Delay` allocates a new timer for every call.

Cool, it was just a suggestion. And I learned something new. Keep the heartbeat.

aik-jahoda commented on src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/Http2KeepAlive.cs:
```diff
@@ -0,0 +1,99 @@
+
+            private readonly TimeSpan _keepAlivePingDelay;
+            private readonly TimeSpan _keepAlivePingTimeout;
+            private DateTimeOffset _nextPingRequestTimestamp;
```
I have removed DateTimeOffset in favor of Environment.TickCount64

aik-jahoda commented on src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/Http2Connection.cs:
```diff
@@ -950,6 +958,21 @@ private async Task ProcessOutgoingFramesAsync()
             });
 
+        /// <param name="pingContent">The 8-byte ping content to send, read as a big-endian integer.</param>
+        private Task SendPingAsync(long pingContent) =>
```
Merged

Commit:
Apply PR comments

Commit:
Merge remote-tracking branch 'upstream/master' into jajahoda/httpping

Commit:
Remove Http2KeepAlice class

aik-jahoda commented on src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/Http2Connection.cs:
```diff
@@ -96,6 +97,8 @@ internal sealed partial class Http2Connection : HttpConnectionBase, IDisposable
         // Channel options for creating _writeChannel
         private static readonly UnboundedChannelOptions s_channelOptions = new UnboundedChannelOptions() { SingleReader = true };
 
+        private Http2KeepAlive _keepAlive;
```
The class was removed

Commit:
fix trace message

Commit:
ProcessPingAck no longer return bool

JamesNK commented on src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/SocketsHttpHandler.cs:
```diff
@@ -280,6 +280,72 @@ public TimeSpan Expect100ContinueTimeout
+                if ((value.Ticks <  TimeSpan.TicksPerSecond && value != Timeout.InfiniteTimeSpan) ||
+                    (value.TotalMilliseconds > int.MaxValue))
+                {
+                    throw new ArgumentOutOfRangeException(nameof(value));
```
For user friendliness I would have an error message here of what the allow range of values are

JamesNK commented on src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/SocketsHttpHandler.cs:
```diff
@@ -280,6 +280,72 @@ public TimeSpan Expect100ContinueTimeout
+                if ((value.Ticks <  TimeSpan.TicksPerSecond && value != Timeout.InfiniteTimeSpan) ||
+                    (value.TotalMilliseconds > int.MaxValue))
+                {
+                    throw new ArgumentOutOfRangeException(nameof(value));
```
For user friendliness I would have an error message here of what the allow range of values are

JamesNK commented on src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/Http2Connection.cs:
```diff
@@ -1849,6 +1890,49 @@ private void RemoveStream(Http2Stream http2Stream)
+            if (Interlocked.Read(ref _keepAlivePingPayload) != payload)
+                ThrowProtocolError();
+            _keepAliveState = KeepAliveState.None;
+            RefreshPingTimestamp();
```
Is `RefreshPingTimestamp` required here? The timestamp should have already been refreshed in `ProcessIncomingFramesAsync`

aik-jahoda replied:
you are right

JamesNK commented on src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/Http2Connection.cs:
```diff
@@ -1849,6 +1890,49 @@ private void RemoveStream(Http2Stream http2Stream)
+                        _keepAliveState = KeepAliveState.PingSent;
+                        _keepAlivePingTimeoutTimestamp = now + (long)_keepAlivePingTimeout.TotalMilliseconds;
+
+                        long pingPayload = Interlocked.Increment(ref _keepAlivePingPayload);
```
Out of interest, is there a reason you want to send each ping with a unique payload and verify it?

aik-jahoda replied:
The reason is just to identify the specific ping ack to ensure server response correctly.

ManickaP reviewed (approved):
Little nits here and there; otherwise LGTM, thanks!

ManickaP reviewed (approved):
Little nits here and there; otherwise LGTM, thanks!

aik-jahoda committed:
Apply suggestions from code review

Co-authored-by: Marie Píchová <11718369+ManickaP@users.noreply.github.com>

Commit:
Add argument out of ramnge message

Commit:
Merge branch 'jajahoda/httpping' of github.com:aik-jahoda/runtime into jajahoda/httpping

aik-jahoda commented on src/libraries/Common/src/System/Net/Http/HttpHandlerDefaults.cs:
```diff
@@ -9,12 +9,12 @@ namespace System.Net.Http
         public const int DefaultMaxConnectionsPerServer = int.MaxValue;
         public const int DefaultMaxResponseDrainSize = 1024 * 1024;
-        public static readonly TimeSpan DefaultResponseDrainTimeout = TimeSpan.FromSeconds(2);
+        public static TimeSpan DefaultResponseDrainTimeout => TimeSpan.FromTicks(TimeSpan.TicksPerSecond * 2);
```
I reverted it back to fields but with tick constructor

aik-jahoda commented on src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/Http2Connection.cs:
```diff
@@ -1849,6 +1890,49 @@ private void RemoveStream(Http2Stream http2Stream)
 
+        private void RefreshPingTimestamp()
+        {
+            _nextPingRequestTimestamp = Environment.TickCount64 + (long)_keepAlivePingDelay.TotalMilliseconds;
```
I was thinking about the same.

aik-jahoda commented on src/libraries/System.Net.Http/tests/FunctionalTests/SocketsHttpHandlerTest.cs:
```diff
@@ -279,6 +279,122 @@ public sealed class SocketsHttpHandler_HttpClientHandler_ServerCertificates_Test
+
+                    // Warmup request to create connection.
+                    await client.GetStringAsync(uri);
+                    //_output.WriteLine("* start req");
```
This should definitely go away

Commit:
Apply PR comments

Commit:
SocketsHttpHandler setters

scalablecory commented on src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/HttpConnectionPoolManager.cs:
```diff
@@ -102,14 +104,30 @@ public HttpConnectionPoolManager(HttpConnectionSettings settings)
+                            {
+                                thisRef.HeartBeat();
+                            }
+                        }, thisRef, TimeSpan.FromSeconds(1), TimeSpan.FromSeconds(1));
```
Running every 1 second seems like too much. How about `max(1000, min(_keepAlivePingDelay, _keepAlivePingTimeout) / 4)`?

scalablecory commented on src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/HttpConnectionPoolManager.cs:
```diff
@@ -517,6 +535,14 @@ private void RemoveStalePools()
 
+        private void HeartBeat()
+        {
+            foreach (HttpConnectionPool pool in _pools.Values)
+            {
+                pool.HeartBeat();
+            }
```
`ConcurrentDictionary.Values` locks the collection and copies all the values to a list -- not good to use here.

Enumerating the full collection acquires no locks.

```suggestion
            foreach (KeyValuePair<HttpConnectionKey, HttpConnectionPool> pool in _pools)
            {
                pool.Value.HeartBeat();
            }
```

Commit:
Merge remote-tracking branch 'upstream/master' into jajahoda/httpping

scalablecory commented on src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/SocketsHttpHandler.cs:
```diff
@@ -280,6 +280,70 @@ public TimeSpan Expect100ContinueTimeout
+                }
+
+                CheckDisposedOrStarted();
+                _settings._keepAlivePingTimeout = value.TotalMilliseconds > int.MaxValue? TimeSpan.FromMilliseconds(int.MaxValue) : value;
```
This would be confusing to the user if they set one value and get another -- can we clamp at a later stage and keep their original value here?

aik-jahoda replied:
done

aik-jahoda committed:
Update src/libraries/Common/tests/System/Net/Http/Http2LoopbackConnection.cs

Co-authored-by: Cory Nelson <phrosty@gmail.com>

aik-jahoda committed:
Update src/libraries/System.Net.Http/src/Resources/Strings.resx

Co-authored-by: Cory Nelson <phrosty@gmail.com>

aik-jahoda committed:
Apply suggestions from code review

Co-authored-by: Cory Nelson <phrosty@gmail.com>

aik-jahoda commented on src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/Http2Connection.cs:
```diff
@@ -1849,6 +1890,48 @@ private void RemoveStream(Http2Stream http2Stream)
+            {
+                case KeepAliveState.None:
+                    // Check whether keep alive delay has passed since last frame received
+                    if (now > _nextPingRequestTimestamp && _keepAlivePingDelay != Timeout.InfiniteTimeSpan.TotalMilliseconds)
```
agree, the timer is disabled in this case

Commit:
PR comments

scalablecory commented on src/libraries/Common/tests/System/Net/Http/Http2LoopbackConnection.cs:
```diff
@@ -631,6 +631,23 @@ public async Task PingPong()
+            Assert.NotNull(frame);
+            Assert.Equal(FrameType.Ping, frame.Type);
+            Assert.Equal(0, frame.StreamId);
+            Assert.False(frame.AckFlag);
```
We should also assert a payload size of 8.

ManickaP replied:
Done.

scalablecory commented on src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/Http2Connection.cs:
```diff
@@ -1849,6 +1896,53 @@ private void RemoveStream(Http2Stream http2Stream)
+                case KeepAliveState.PingSent:
+                    if (now > _keepAlivePingTimeoutTimestamp)
+                        ThrowProtocolError();
+                    break;
```
A `default: Debug.Fail()` would be nice here.

ManickaP replied:
Done.

scalablecory commented on src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/HttpConnectionPool.cs:
```diff
@@ -1689,6 +1689,17 @@ private static bool GetIsWindows7Or2008R2()
+            if (localHttp2Connections != null)
+            {
+                foreach (Http2Connection http2Connection in localHttp2Connections)
+                    http2Connection?.HeartBeat();
```
@alnikola will this collection ever have a null connection in it?

```suggestion
                    http2Connection.HeartBeat();
```

ManickaP replied:
Fixed, AFAIK this will never be `null`, the array gets resized with each addition/removal.

scalablecory reviewed (commented):
Looking good.

ManickaP committed:
Merge branch 'master' into jajahoda/httpping

ManickaP committed:
Addressed the last batch of PR feedback.

ManickaP commented:
/azp run runtime-libraries outerloop

ManickaP committed:
Merge branch 'master' into jajahoda/httpping

ManickaP committed:
Fixed KeepAlive test dependency on frame chronology.

On slower platforms, PINGs were coming sooner than requests themselves causing errors while reading the request and receiving PING frame instead of HEADERS frame.

ManickaP commented:
/azp run runtime-libraries outerloop

ManickaP committed:
Fixed FW compilation.

ManickaP commented:
/azp run runtime-libraries outerloop

ManickaP committed:
Split and moved tests to theirs proper places.

ManickaP commented:
/azp run runtime-libraries outerloop

ManickaP committed:
More test fixing for super slow CI.

ManickaP commented:
/azp run runtime-libraries outerloop

danmosemsft commented on src/libraries/System.Net.Http/src/Resources/Strings.resx:
```diff
@@ -264,6 +264,9 @@
     <value>The specified value must be greater than {0}.</value>
   </data>
+  <data name="net_http_value_must_be_greater_than_or_equal" xml:space="preserve">
+    <value>The specified value must be greater than or equal {0}.</value>
```
```suggestion
    <value>The specified value must be greater than or equal to {0}.</value>
```
would it be useful to include the "specified value" ? Sometimes exception messages are all you have out of a log.

ManickaP replied:
Fixed.

danmosemsft commented on src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/SocketsHttpHandler.cs:
```diff
@@ -280,6 +280,70 @@ public TimeSpan Expect100ContinueTimeout
+            {
+                if (value.Ticks < TimeSpan.TicksPerSecond && value != Timeout.InfiniteTimeSpan)
+                {
+                    throw new ArgumentOutOfRangeException(nameof(value), value, SR.Format(SR.net_http_value_must_be_greater_than_or_equal, "1 second"));
```
Although we dont localize our exception messages today, we may in future, and when we do that ideally we won't need code changes. Unfortunately if we create messages by concatenating sentence fragments it is not localizable. Eg., consider if we must localize such that "second" has to grammatically agree with the rest of the sentence, or "1" and "second" must not be next to each other. Possible approaches
1. "The specified value must be greater than or equal to {0} seconds." 
1. "The specified value must be greater than or equal to &quot;{0}&quot;"

I think #1 is what we usually do.

ManickaP replied:
Fixed.

ManickaP committed:
Exception texts.

MihaZupan commented:
@ManickaP FYI The `Http2_PingKeepAlive` outerloop test seems to be [failing quite a bit](https://dataexplorer.azure.com/clusters/engsrvprod/databases/engineeringdata?query=H4sIAAAAAAAAA22OvQrCQBCEe59isVFBIaayiaCFGEVRESzlzA1mNdkLdxttfHjjD1ZWs8N87MweQXcIdaGBWg+6OBa6stiEReDp4Pw1VZSBnPxMav+hC3d6U42mtsnvOTxoxsIhh6VxQtYolEt04yiOBtFoEA17LTJiaQXNnaUkofZctYqPG5bzEqgmBd/Q/kCfmS+oMzNcdJqOyrsLMv219L+fXhqCOaNP2xo11qZszqk3kuVPYNToZPIAAAA=)

karelz commented:
Is there a bug tracking the failure?

ManickaP commented:
It's bug in test, no need for tracking issue, I'm working on it right know.

aik-jahoda commented:
@ManickaP, thanks for finish this PR!
